### PR TITLE
chore: refactor load data workflow

### PR DESCRIPTION
**Pull request by amliuyong**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

(describe what this merge request does)

## Implementation highlights

(describe how the merge request does for feature changes, share the RFC link if it has)

## Test checklist

- [x] add new test cases
- [ ] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [x] deploy data modeling
    - [x] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/analytics/lambdas/load-data-workflow/has-more-job-new.ts [hunk 1]
```diff
@@ -30,28 +32,55 @@ export const handler = async (_: any, context: Context) => {
...
-    jobNewCount += newRecordResp.Count;
-    if (newRecordResp.LastEvaluatedKey) {
-      lastEvaluatedKey = newRecordResp.LastEvaluatedKey;
-    } else {
-      break;
+  const getStatusFilesCount = async (redshiftTableName: string, jobStatus: string) => {
+
+    let lastEvaluatedKey = undefined;
+    const jobStatusQuery = composeJobStatus(jobStatus, redshiftTableName);
+    const prefixQuery = odsEventBucketWithPrefix.replace(new RegExp(`\/${REDSHIFT_ODS_TABLE_NAME}\/?$`), `/${redshiftTableName}/`);
```
[the remark this answers is not in the document]
Comment (amliuyong) [reply src/analytics/lambdas/load-data-workflow/has-more-job-new.ts:42]: changed to `new RegExp(`/${REDSHIFT_ODS_TABLE_NAME}/?$`)`, 
"\/" is unnecessary.  
check the sonarqube checks.

File: src/analytics/lambdas/load-data-workflow/create-load-manifest.ts [hunk 2]
```diff
@@ -131,7 +131,17 @@ export const handler = async (_event: any, context: Context) => {
     }
   }
 
-  logger.info(`allJobNewCount: ${allJobNewCount}`);
+  logger.info(`allJobNewCount: ${allJobNewCount}, candidateItems.length: ${candidateItems.length}`);
+
+  // reprocess files in JOB_PROCESSING
+  let processingRecordResp = await queryItems(tableName, indexName, odsEventBucketWithPrefix, JobStatus.JOB_PROCESSING, undefined);
```
[the remark this answers is not in the document]
Comment (amliuyong) [reply src/analytics/lambdas/load-data-workflow/create-load-manifest.ts:137]: the step `Has Other Running Workflow` can make sure only one step function is in running state.
 
if we see JOB_PROCESSING files here, they are failed-loaded files by previous step function, here add them to candidate list and reprocess them.

File: src/analytics/private/load-ods-data-workflow.ts [hunk 3]
```diff
@@ -62,47 +69,40 @@ export class LoadOdsDataToRedshiftWorkflow extends Construct {
...
-          },
-          object: {
-            key: Match.prefix(props.odsSource.prefix),
-          },
-        },
-        source: ['aws.s3'],
-      },
-    });
-    sourceTriggerRule.addTarget(new LambdaFunction(processorLambda));
+      const odsSource: ODSSource = (props.tablesOdsSource as any)[`${tableName}`];
```
[the remark this answers is not in the document]
Comment (amliuyong) [reply src/analytics/private/load-ods-data-workflow.ts:74]: changed to ```
const [tableName, odsSource] of Object.entries(props.tablesOdsSource)
```

File: src/analytics/lambdas/load-data-workflow/create-load-manifest.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (zxkane) [reply src/analytics/lambdas/load-data-workflow/create-load-manifest.ts:137]: What's your mechanism to exit the infinite loop if some files are unable to be loaded by Redshift? Do you have a way like DLQ to handle this case?

[the remark this answers is not in the document]
Comment (amliuyong) [reply src/analytics/lambdas/load-data-workflow/create-load-manifest.ts:137]: If there are files in New or Processing status, the workflow will continue to run util loading them all. 
If files are unable to be loaded(e.g. bad files, wrong schema), the workflow will continue failing and exit, alarm will be raised. In this situation, we need to manually check.

**Review by zxkane — APPROVED** [review]

**Pull request merged**